src/grafter_2/rdf/protocols.cljc

```diff
@@ -0,0 +1,616 @@
+
+     IRawValue
+     (raw-value [this]
+       (.stringValue this))
+
+     IDatatypeURI
+     (datatype-uri [this]
+       (URI. (str (.getDatatype this))))))
+
+(defrecord RDFLiteral [raw-value datatype-uri]
+  IRawValue
+  (raw-value [this]
+    (:raw-value this))
+
+  IDatatypeURI
+  (datatype-uri [this]
+    (:datatype-uri this))
+
+  IRDFString
+  (lang [this]
+    nil))
+
+(defn literal
+  "You can use this to declare an RDF typed literal value along with
+  its URI.  Note that there are implicit coercions already defined for
+  many core clojure/java datatypes, so for common datatypes you
+  shouldn't need this."
+  [val datatype-uri]
+  (->RDFLiteral (str val)
+                #?(:clj  (->java-uri datatype-uri)
+                   :cljs (->cljs-uri datatype-uri))))
+
+(extend-protocol IRawValue
+  nil
+  (raw-value [t]
+    t))
+
+#?(:clj
+   (extend-protocol IRawValue
+     Object
+     (raw-value [t]
+       t))
+   :cljs
+   (extend-protocol IRawValue
+     object
+     (raw-value [t]
+       t)))
+
+#?(:cljs
+   (extend-protocol IDatatypeURI
+     boolean
+     (datatype-uri [t] gvx/xsd:boolean)
+
+     string
+     (datatype-uri [t] gvx/xsd:string))
+
+   :clj
+   (extend-protocol IDatatypeURI
+     Boolean
+     (datatype-uri [t] xsd:boolean)
+
+     String
+     (datatype-uri [t] xsd:string)
+
+     java.math.BigInteger
+     (datatype-uri [t]
+       (->java-uri xsd:integer))
+
+     clojure.lang.BigInt
+     (datatype-uri [t]
+       (->java-uri xsd:integer))
+
+     java.math.BigDecimal
+     (datatype-uri [t]
+       (->java-uri xsd:decimal))
+
+     Byte
+     (datatype-uri [t]
+       (->java-uri xsd:byte))
+
+     LocalTime
+     (datatype-uri [t]
+       (->java-uri xsd:time))
+
+     OffsetTime
+     (datatype-uri [t]
+       (->java-uri xsd:time))
+
+     LocalDate
+     (datatype-uri [t]
+       (->java-uri xsd:date))
+
+     OffsetDate
+     (datatype-uri [t]
+       (->java-uri xsd:date))
+
+     LocalDateTime
+     (datatype-uri [t]
+       (->java-uri xsd:dateTime))
+
+     OffsetDateTime
+     (datatype-uri [t]
+       (->java-uri xsd:dateTime))
+
+     Double
+     (datatype-uri [t]
+       (->java-uri xsd:double))
+
+     Float
+     (datatype-uri [t]
+       (->java-uri xsd:float))
+
+     Integer
+     (datatype-uri [t]
+       (->java-uri xsd:integer))
+
+     Long
+     (datatype-uri [t]
+       (->java-uri xsd:integer))
+
+     Short
+     (datatype-uri [t]
+       (->java-uri xsd:short))))
+
+(defn- destructure-quad [quad i default]
+  (case i
+    0 (:s quad)
+    1 (:p quad)
+    2 (:o quad)
+    3 (or (:c quad) default)
+    :else default))
+
+#?(:clj
+   (defrecord Quad [s p o c]
+
+     IStatement
+     (subject [s] (:s s))
+     (predicate [s] (:p s))
+     (object [s] (:o s))
+     (context [s] (:c s))
+
+     clojure.lang.Indexed
+     (nth [this ^int i]
+       (destructure-quad this i nil))
+
+     (nth [this ^int i default]
+       (destructure-quad this i default)))
+
```

Comment:
Because `clojure.lang.Indexed` is an interface and not a protocol, I could not properly extend to it outside of the defrecord macro, so I ended up repeating myself with the defrecord form and IStatement.